app/controllers/application_controller.rb

```diff
@@ -5,6 +5,144 @@ class ApplicationController < Sinatra::Base
   configure do
     set :public_folder, 'public'
     set :views, 'app/views'
+    enable :sessions
+    set :session_secret, 'totally_secure_key'
   end
 
-end
\ No newline at end of file
+  get "/" do
+    if logged_in?
+      redirect "/tweets"
+    else
+      erb :index
+    end
+  end
+
+  get "/signup" do
+    if logged_in?
+      redirect "/tweets"
+    else
+      erb :signup
+    end
+  end
+
+  post "/signup" do
+    user = User.new(params)
+    if user.save
+      session[:user_id] = user.id
+      redirect "/tweets"
+    else
+      redirect "/signup"
+    end
+  end
+
+  get "/login" do
+    if logged_in?
+      redirect "/tweets"
+    else
+      erb :login
+    end
+  end
+
+  post "/login" do
+    user = User.find_by(username: params[:username])
+    if user && user.authenticate(params[:password])
+      session[:user_id] = user.id
+      redirect "/tweets"
+    else
+      redirect "/login"
+    end
+  end
+
+  get "/logout" do
+    session.clear
+    redirect "/login"
+  end
+
+  get "/tweets" do
+    as_logged_in do
+      @tweets = Tweet.all
+      erb :'tweets/index'
+    end
+  end
+
+  post "/tweets" do
+    as_logged_in do |user|
+      tweet = Tweet.new(content: params[:content], user: user)
+      if tweet.save
+        redirect "/tweets"
+      else
+        redirect "/tweets/new"
+      end
+    end
+  end
+
+  get "/tweets/new" do
+    as_logged_in do
+      erb :'tweets/new'
+    end
+  end
+
+  get "/tweets/:id" do
+    as_logged_in do |user|
+      @tweet = Tweet.find(params[:id])
+      erb :'tweets/show'
+    end
+  end
+
+  patch "/tweets/:id" do
+    as_logged_in do |user|
+      @tweet = Tweet.find(params[:id])
+      if user==@tweet.user && @tweet.update(content: params[:content])
+        redirect "/tweets/#{params[:id]}"
+      else
+        redirect "/tweets/#{params[:id]}/edit"
+      end
+    end
+  end
+
+  delete "/tweets/:id" do
+    as_logged_in do |user|
+      @tweet = Tweet.find(params[:id])
```

Comment:
Prefer to find tweets that are already only associated to the `current_user` than to manually check the `user_id` foreign key of the tweet against the `current_user.id`

```ruby
tweet = current_user.tweets.find_by(id: params[:id])
if tweet && tweet.destroy
  redirect "/tweets"
else
  redirect "/tweets/#{tweet.id}"
end
```

We're using the association collection method provided by [`has_many`](http://api.rubyonrails.org/classes/ActiveRecord/Associations/ClassMethods.html#method-i-has_many) to query the user's tweets. 

Note the usage of [`find_by`](http://api.rubyonrails.org/classes/ActiveRecord/FinderMethods.html#method-i-find-by) vs [`find`](http://api.rubyonrails.org/classes/ActiveRecord/FinderMethods.html#method-i-find). Had we used `find` here, ActiveRecord would raise an error in the event that we could not find a tweet associated with this user. `find_by` will allow this action to fail gracefully, simply returning nil if the user is trying to delete a tweet that they do not have access to.